cmd/certificate-generator/app/certificate-generator.go

```diff
@@ -0,0 +1,313 @@
+	serverKey, err := ecdsa.GenerateKey(elliptic.P384(), rand.Reader)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	dnsNames := make([]string, 2*len(*opts.WebhookNames))
+	for idx := -1; idx < len(*opts.WebhookNames)-1; idx++ {
+		dnsNames[idx+1] = fmt.Sprintf("%v.%v.svc", (*opts.WebhookNames)[idx+1], opts.WebhookNamespace)
+		dnsNames[idx+2] = fmt.Sprintf("%v.%v.svc.%v", (*opts.WebhookNames)[idx+1], opts.WebhookNamespace, opts.KubernetesDomain)
+	}
+
+	serverTemplate := x509.Certificate{
+		SerialNumber: big.NewInt(2),
+		Subject: pkix.Name{
+			Country:    []string{"US"},
+			CommonName: opts.WebhookServerName,
+		},
+		NotBefore:   time.Now(),
+		NotAfter:    time.Now().Add(time.Duration(opts.CertValidity) * time.Minute),
+		KeyUsage:    x509.KeyUsageDigitalSignature,
+		ExtKeyUsage: []x509.ExtKeyUsage{x509.ExtKeyUsageServerAuth},
+		DNSNames:    dnsNames,
+	}
+
+	caBlock, _ := pem.Decode(caCert)
+	ca, err := x509.ParseCertificate(caBlock.Bytes)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	serverCert, err := x509.CreateCertificate(rand.Reader, &serverTemplate, ca, &serverKey.PublicKey, caKey)
+	if err != nil {
+		return nil, nil, err
+	}
+
+	certPem := pem.EncodeToMemory(&pem.Block{Type: "CERTIFICATE", Bytes: serverCert})
+	marshalledKey, err := x509.MarshalECPrivateKey(serverKey)
+	if err != nil {
+		return nil, nil, fmt.Errorf("error marshalling webhook server certificate key: %v", err)
+	}
+
+	keyPem := pem.EncodeToMemory(&pem.Block{Type: "EC PRIVATE KEY", Bytes: marshalledKey})
+
+	return certPem, keyPem, nil
+}
+
+func updateSecret(clientset *kubernetes.Clientset, certPEM, keyPEM, caPEM []byte, opts *options.Options) error {
+	secret := &corev1.Secret{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      opts.SecretName,
+			Namespace: opts.SecretNamespace,
+		},
+		Type: corev1.SecretTypeTLS,
+		Data: map[string][]byte{
+			"tls.crt": certPEM,
+			"tls.key": keyPEM,
+			"ca.crt":  caPEM,
+		},
+	}
+
+	_, err := clientset.CoreV1().Secrets(opts.SecretNamespace).Update(context.TODO(), secret, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update secret: %v", err)
+	}
+
+	return nil
+}
+
+// Update the webhook configuration with the new CA bundle
+func updateWebhookConfiguration(client *kubernetes.Clientset, opts *options.Options, caCert []byte) error {
+	webhook, err := client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Get(context.Background(), opts.WebhookServerName, metav1.GetOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to get webhook configuration: %v", err)
+	}
+
+	for i := range webhook.Webhooks {
+		webhook.Webhooks[i].ClientConfig.CABundle = caCert
+	}
+
+	_, err = client.AdmissionregistrationV1().ValidatingWebhookConfigurations().Update(context.Background(), webhook, metav1.UpdateOptions{})
+	if err != nil {
+		return fmt.Errorf("failed to update webhook configuration: %v", err)
+	}
+
+	return nil
+}
```

Comment:
Few questions:
1.  How does webhook server load the rotated cert file
2.  How does webhook server hot reload the cert file
3.  How to sync from secret to local dir

Reply:
1. How does webhook server load the rotated cert file
    A. The webhook server will get the certificate from the local dir when in starts using the code below:
    https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.4/pkg/webhook/server.go#L201-L203

2. How does webhook server hot reload the cert file
    A. The webhook server automatically starts a cert-watcher as a background go routine when it starts. This cert-watcher will watch the certificates for changes and reload them when necessary.
    https://github.com/kubernetes-sigs/controller-runtime/blob/v0.18.4/pkg/webhook/server.go#L207-L217

3. How to sync from secret to local dir
    A. In kubernetes the local dir will automatically reflect the changes in the secret since the secret is mounted there during pod creation. No need for manual sync.